sdk/core/azure-core/src/main/java/com/azure/core/util/AuthorizationChallengeHandler.java

```diff
@@ -0,0 +1,521 @@
+
+    /**
+     * Header representing additional information a proxy server is expecting during future authentication requests.
+     */
+    public static final String PROXY_AUTHENTICATION_INFO = "Proxy-Authentication-Info";
+
+    private final String username;
+    private final String password;
+    private final Map<String, AtomicInteger> nonceTracker;
+    private final AtomicReference<String> authorizationPipeliningType = new AtomicReference<>();
+    private final AtomicReference<Map<String, String>> lastChallenge = new AtomicReference<>();
+
+    /**
+     * Creates an {@link AuthorizationChallengeHandler} using the {@code username} and {@code password} to respond to
+     * authentication challenges.
+     *
+     * @param username Username used to response to authorization challenges.
+     * @param password Password used to respond to authorization challenges.
+     */
+    public AuthorizationChallengeHandler(String username, String password) {
+        this.username = username;
+        this.password = password;
+        this.nonceTracker = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Handles Basic authentication challenges.
+     *
+     * @return Authorization header for Basic authentication challenges.
+     */
+    public final String handleBasic() {
+        authorizationPipeliningType.set(BASIC);
+        String token = username + ":" + password;
+        return BASIC + Base64.getEncoder().encodeToString(token.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Handles Digest authentication challenges.
+     *
+     * @param method HTTP method being used in the request.
+     * @param uri Relative URI for the request.
+     * @param challenges List of challenges that the server returned for the client to choose from and use when creating
+     * the authorization header.
+     * @param entityBodySupplier Supplies the request entity body, used to compute the hash of the body when using
+     * {@code "qop=auth-int"}.
+     * @return Authorization header for Digest authentication challenges.
+     */
+    public final String handleDigest(String method, String uri, List<Map<String, String>> challenges,
+        Supplier<byte[]> entityBodySupplier) {
+        authorizationPipeliningType.set(DIGEST);
+        Map<String, List<Map<String, String>>> challengesByType = partitionByChallengeType(challenges);
+
+        for (String algorithm : ALGORITHM_PREFERENCE_ORDER) {
+            // No challenges using this algorithm, skip it.
+            if (!challengesByType.containsKey(algorithm)) {
+                continue;
+            }
+
+            Function<byte[], byte[]> digestFunction = getDigestFunction(algorithm);
+
+            // Unable to retrieve a digest for the specified algorithm, skip it.
+            if (digestFunction == null) {
+                continue;
+            }
+
+            Map<String, String> challenge = new ConcurrentHashMap<>(challengesByType.get(algorithm).get(0));
```

Comment:
Is it guaranteed that the first entry in the list will be the latest?

Reply:
It is guaranteed, this code is a bit odd to cover some potential corner cases where multiple challenges using the same algorithm are returned by a proxy server though it is unlikely.